### DISCUSSION TITLE
url_for: Allow overriding default values

### DISCUSSION THREAD

**Discussion by NickVeld** [pallets/werkzeug#2456] [Q&A] [answered]
Sometimes defining a route the default path values can be set for the user convenience.
In the same time several usage example can be needed to show on a "help" page.
The most popular way to generate a URL is `url_for`.
However, it seems that it is impossible to generate the URL with the path arguments having the default value.

Firstly, I have asked a [question](https://stackoverflow.com/q/73028511/8854890) about it on StackOverflow. But no other way to override has been introduced.

(It is a simplified example that reproduce the problem.)

I have `@app.route("/api/<a>/<b>", defaults={"a": "A0", "b": "B0"})`

I want to generate the following URL: `/api/A0/B1` (in other word, with the overridden argument `b`).

The most obvious to me approach: `flask.url_for(route.endpoint, **{"b": "B1"})`.
It causes `werkzeug.routing.BuildError: Could not build url for endpoint 'SOME_ENDPOINT' with values ['b']. Did you forget to specify values ['a']?`

The next approach: `flask.url_for(route.endpoint, **{"a": "A0", "b": "B1"})`.
It causes `werkzeug.routing.BuildError: Could not build url for endpoint 'index.API.event_average_time_in_seconds' with values ['a', 'b'].`. It is triggered by the following code from `werkzeug.routing` with my comments after `<--`:
```
        # in case defaults are given we ensure that either the value was
        # skipped or the value is the same as the default value. <-- why it is needed anyway??
        if defaults:
            for key, value in defaults.items():
                if key in values and value != values[key]: # <-- of course, it is false
                    return False  # <-- it returns `False` that used to trigger the `BuildError` above
```

The next approach: `route.defaults['b'] = 'B1'; flask.url_for(route.endpoint, **route.defaults)`. It returns the following URL somewhy: `/api/A0/B0`. In other words, no exception but the generated URL uses the defaults that was not overrided. My debugger hangs on the URL building step so I do not know why the system checks against overridden `defaults` but generates URL with the initial one.

I have also tried the approach from [this answer](https://stackoverflow.com/a/33783451) but `@app.url_defaults` behaves in the same way as `flask.url_for(..., **{...})` in terms of this problem.

Firstly, I have tested it on `Python==3.8.13 Flask==2.0.3 Werkzeug==2.1.2`, then on `Python==3.8.13 Flask==2.1.3 Werkzeug==2.1.2`

The only way not changing Flask I come up with is `generated_url.replace('B0', 'B1')` (the string constants are from the situation above)

**Comment by davidism** [comment]
It will be possible to override the behavior of `url_for` in Flask 2.2 with #4580. If you don't want to write rules to cover each variable default as shown in [my other answer](https://github.com/pallets/werkzeug/discussions/2456#discussioncomment-3198384), you could override `app.url_for` to detect the endpoint and inject missing values.

```python
class MyFlask(Flask):
    def url_for(self, endpoint, **kwargs):
        if endpoint == "api_items":
            kwargs.setdefault("a", "A0")
            kwargs.setdefault("b", "B0")

        return super().url_for(endpoint, **kwargs)
```

This is _probably_ not a good idea, it's likely less efficient to check every call to see if it's for one endpoint than to use multiple rules and let the router take care of it.

**Reply by NickVeld** [reply]
I agree that it is probably not a good idea. I would rather go with proper definition from another answer.

**Comment by davidism** [comment, marked as answer]
This comes from a misunderstanding of how `defaults` works in Werkzeug's router. From the docs for `werkzeug.routing.Rule`:

```
`defaults`
    An optional dict with defaults for other rules with the same endpoint.
    This is a bit tricky but useful if you want to have unique URLs::

        url_map = Map([
            Rule('/all/', defaults={'page': 1}, endpoint='all_entries'),
            Rule('/all/page/<int:page>', endpoint='all_entries')
        ])

    If a user now visits ``http://example.com/all/page/1`` they will be
    redirected to ``http://example.com/all/``.  If `redirect_defaults` is
    disabled on the `Map` instance this will only affect the URL
    generation.
```

So you would use defaults to create a default URL that does not have those corresponding variables, and it would be routed to the correct endpoint with those defaults. It will also generate the short URL if the default values are given, and will redirect to the short URL if the default URL is given. It will _not_ fill in missing values, or allow overriding the variables of a single rule. It requires both a Rule with the defaults (and no variables), and a Rule with the variables.

Here's how you could use four rules to allow the behavior you want, overriding any combination of the variables. Typically though, you wouldn't write the default values in the URL still, you'd do what the example in the docs above show and have a default rule with no values in the URL.

```python
@app.route("/api/A0/B0", defaults={"a": "A0", "b": "B0"})
@app.route("/api/<a>/B0", defaults={"b": "B0"})
@app.route("/api/A0/<b>", defaults={"a": "A0"})
@app.route("/api/<a>/<b>")
def api_items(a, b):
    return f"{a=}, {b=}"
```

**Reply by davidism** [reply]
Posted this with some more edits as an answer to the linked question: https://stackoverflow.com/a/73067944/400617

**Reply by NickVeld** [reply]
Separate thanks for the extensive explanation and example!

**Reply by NickVeld** [reply]
Is there no way to introduce some other `defaults` argument that reduces the number of `api.route` definition for my case to one? Or some separate `api.route` generator?
I can do it myself for me. But I suppose it can useful in the package.

**Reply by NickVeld** [reply]
Only my task solves the following code:
```
@app.route("/api", defaults={"a": "A0", "b": "B0"})
@app.route("/api/<a>", defaults={"b": "B0"})
@app.route("/api/<a>/<b>")
def api_items(a, b):
    return f"{a=}, {b=}"
```

I suppose you mean it too. Am I right?

`@app.route("/api/<b>", defaults={"a": "A0"})` must be invalid because for the rule parser it is the same as `@app.route("/api/<a>", defaults={"b": "B0"})`
